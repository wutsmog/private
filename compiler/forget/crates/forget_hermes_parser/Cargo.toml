[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }