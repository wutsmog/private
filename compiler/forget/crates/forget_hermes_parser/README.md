# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.