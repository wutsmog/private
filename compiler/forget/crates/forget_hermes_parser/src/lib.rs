mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
