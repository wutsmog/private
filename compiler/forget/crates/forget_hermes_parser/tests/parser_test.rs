use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
