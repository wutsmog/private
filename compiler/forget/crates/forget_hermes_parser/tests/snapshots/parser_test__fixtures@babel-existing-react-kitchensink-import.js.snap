---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/babel-existing-react-kitchensink-import.js
---
Input:
import * as React from "react";
import { useState, useMemo } from "react";

function Component(props) {
  const [x] = useState(0);
  const expensiveNumber = useMemo(() => calculateExpensiveNumber(x), [x]);

  return <div>{expensiveNumber}</div>;
}

function Component2(props) {
  const [x] = useState(0);
  const expensiveNumber = useMemo(() => calculateExpensiveNumber(x), [x]);

  return <div>{expensiveNumber}</div>;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "ImportDeclaration",
      "specifiers": [
        {
          "type": "ImportNamespaceSpecifier",
          "local": {
            "type": "Identifier",
            "name": "React",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 12,
              "end": 17
            }
          },
          "loc": null,
          "range": {
            "start": 7,
            "end": 17
          }
        }
      ],
      "source": {
        "type": "StringLiteral",
        "value": "react",
        "loc": null,
        "range": {
          "start": 23,
          "end": 30
        }
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 31
      }
    },
    {
      "type": "ImportDeclaration",
      "specifiers": [
        {
          "type": "ImportSpecifier",
          "imported": {
            "type": "Identifier",
            "name": "useState",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 41,
              "end": 49
            }
          },
          "local": {
            "type": "Identifier",
            "name": "useState",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 41,
              "end": 49
            }
          },
          "loc": null,
          "range": {
            "start": 41,
            "end": 49
          }
        },
        {
          "type": "ImportSpecifier",
          "imported": {
            "type": "Identifier",
            "name": "useMemo",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 51,
              "end": 58
            }
          },
          "local": {
            "type": "Identifier",
            "name": "useMemo",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 51,
              "end": 58
            }
          },
          "loc": null,
          "range": {
            "start": 51,
            "end": 58
          }
        }
      ],
      "source": {
        "type": "StringLiteral",
        "value": "react",
        "loc": null,
        "range": {
          "start": 66,
          "end": 73
        }
      },
      "loc": null,
      "range": {
        "start": 32,
        "end": 74
      }
    },
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 85,
          "end": 94
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 95,
            "end": 100
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 113,
                        "end": 114
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 112,
                    "end": 115
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useState",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 118,
                      "end": 126
                    }
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 0.0,
                      "loc": null,
                      "range": {
                        "start": 127,
                        "end": 128
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 118,
                    "end": 129
                  }
                },
                "loc": null,
                "range": {
                  "start": 112,
                  "end": 129
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 106,
              "end": 130
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "expensiveNumber",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 139,
                    "end": 154
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useMemo",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 157,
                      "end": 164
                    }
                  },
                  "arguments": [
                    {
                      "type": "ArrowFunctionExpression",
                      "id": null,
                      "params": [],
                      "body": {
                        "type": "CallExpression",
                        "callee": {
                          "type": "Identifier",
                          "name": "calculateExpensiveNumber",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 171,
                            "end": 195
                          }
                        },
                        "arguments": [
                          {
                            "type": "Identifier",
                            "name": "x",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 196,
                              "end": 197
                            }
                          }
                        ],
                        "loc": null,
                        "range": {
                          "start": 171,
                          "end": 198
                        }
                      },
                      "generator": true,
                      "async": false,
                      "loc": null,
                      "range": {
                        "start": 165,
                        "end": 198
                      },
                      "expression": true,
                      "loc": null,
                      "range": {
                        "start": 165,
                        "end": 198
                      }
                    },
                    {
                      "type": "ArrayExpression",
                      "elements": [
                        {
                          "type": "Identifier",
                          "name": "x",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 201,
                            "end": 202
                          }
                        }
                      ],
                      "loc": null,
                      "range": {
                        "start": 200,
                        "end": 203
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 157,
                    "end": 204
                  }
                },
                "loc": null,
                "range": {
                  "start": 139,
                  "end": 204
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 133,
              "end": 205
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 217,
                    "end": 220
                  }
                },
                "attributes": [],
                "selfClosing": false,
                "loc": null,
                "range": {
                  "start": 216,
                  "end": 221
                }
              },
              "children": [
                {
                  "type": "JSXExpressionContainer",
                  "expression": {
                    "type": "Identifier",
                    "name": "expensiveNumber",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 222,
                      "end": 237
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 221,
                    "end": 238
                  }
                }
              ],
              "closingElement": {
                "type": "JSXClosingElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 240,
                    "end": 243
                  }
                },
                "loc": null,
                "range": {
                  "start": 238,
                  "end": 244
                }
              },
              "loc": null,
              "range": {
                "start": 216,
                "end": 244
              }
            },
            "loc": null,
            "range": {
              "start": 209,
              "end": 245
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 102,
          "end": 247
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 76,
        "end": 247
      },
      "loc": null,
      "range": {
        "start": 76,
        "end": 247
      }
    },
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component2",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 258,
          "end": 268
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 269,
            "end": 274
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 287,
                        "end": 288
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 286,
                    "end": 289
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useState",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 292,
                      "end": 300
                    }
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 0.0,
                      "loc": null,
                      "range": {
                        "start": 301,
                        "end": 302
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 292,
                    "end": 303
                  }
                },
                "loc": null,
                "range": {
                  "start": 286,
                  "end": 303
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 280,
              "end": 304
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "expensiveNumber",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 313,
                    "end": 328
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useMemo",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 331,
                      "end": 338
                    }
                  },
                  "arguments": [
                    {
                      "type": "ArrowFunctionExpression",
                      "id": null,
                      "params": [],
                      "body": {
                        "type": "CallExpression",
                        "callee": {
                          "type": "Identifier",
                          "name": "calculateExpensiveNumber",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 345,
                            "end": 369
                          }
                        },
                        "arguments": [
                          {
                            "type": "Identifier",
                            "name": "x",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 370,
                              "end": 371
                            }
                          }
                        ],
                        "loc": null,
                        "range": {
                          "start": 345,
                          "end": 372
                        }
                      },
                      "generator": true,
                      "async": false,
                      "loc": null,
                      "range": {
                        "start": 339,
                        "end": 372
                      },
                      "expression": true,
                      "loc": null,
                      "range": {
                        "start": 339,
                        "end": 372
                      }
                    },
                    {
                      "type": "ArrayExpression",
                      "elements": [
                        {
                          "type": "Identifier",
                          "name": "x",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 375,
                            "end": 376
                          }
                        }
                      ],
                      "loc": null,
                      "range": {
                        "start": 374,
                        "end": 377
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 331,
                    "end": 378
                  }
                },
                "loc": null,
                "range": {
                  "start": 313,
                  "end": 378
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 307,
              "end": 379
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 391,
                    "end": 394
                  }
                },
                "attributes": [],
                "selfClosing": false,
                "loc": null,
                "range": {
                  "start": 390,
                  "end": 395
                }
              },
              "children": [
                {
                  "type": "JSXExpressionContainer",
                  "expression": {
                    "type": "Identifier",
                    "name": "expensiveNumber",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 396,
                      "end": 411
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 395,
                    "end": 412
                  }
                }
              ],
              "closingElement": {
                "type": "JSXClosingElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 414,
                    "end": 417
                  }
                },
                "loc": null,
                "range": {
                  "start": 412,
                  "end": 418
                }
              },
              "loc": null,
              "range": {
                "start": 390,
                "end": 418
              }
            },
            "loc": null,
            "range": {
              "start": 383,
              "end": 419
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 276,
          "end": 421
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 249,
        "end": 421
      },
      "loc": null,
      "range": {
        "start": 249,
        "end": 421
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 421
  }
}
